typescript  programmatic use (#980)
strapi/documentation: docs/developer-docs/latest/development/typescript.md

3. Choose `runLifecyclesfunction` from the list.
4. When the `strapi.runLifecyclesFunctions` method is added, a list of available lifecycle types (i.e. `register`, `bootstrap` and `destroy`) are returned by the code editor. Use keyboard arrows to choose one of the lifecycles and the code will autocomplete.

## Generate typings for project schemas

> To generate typings for your project schemas use the Strapi CLI and the command `ts:generate-types`. The `ts:generate-types` command creates the file `schemas.d.ts`, at the project root, which stores the schema typings. The optional `--verbose` flag returns the project type definitions in the terminal.

Reviewer: ```suggestion
To generate typings for your project schemas use the [`ts:generate-types` CLI command](/developer-docs/latest/developer-resources/cli/CLI.md#strapi-ts-generate-types). The `ts:generate-types` command creates the file `schemas.d.ts`, at the project root, which stores the schema typings. The optional `--verbose` flag returns the project type definitions in the terminal.
```

> To generate typings for your project schemas use the Strapi CLI and the command `content-types:generate-types`. The `content-types:generate-types` command creates the file `schemas.d.ts`, at the project root, which stores the schema typings. The optional `--verbose` flag returns the project type definitions in the terminal.

Reviewer 2: `ts:generate-types` :)

Reviewer 2: Also, the list of flags and what they do is explained [here](https://github.com/strapi/strapi/pull/13460)

![image](https://user-images.githubusercontent.com/25851739/180231470-504761d6-a085-422e-a5ba-e1321a2d8338.png)

`--silent` completely removes all the logs in the terminal
`--verbose` display a detailed table of the generated schemas (but not the exact definitions)

> To generate typings for your project schemas use the [`ts:generate-types` CLI command](/developer-docs/latest/developer-resources/cli/CLI.md#strapi-ts-generate-types). The `ts:generate-types` command creates the file `schemas.d.ts`, at the project root, which stores the schema typings. The optional `--verbose` flag returns the project type definitions in the terminal.

Reviewer 2: ```suggestion
To generate typings for your project schemas use the [`ts:generate-types` CLI command](/developer-docs/latest/developer-resources/cli/CLI.md#strapi-ts-generate-types). The `ts:generate-types` command creates the file `schemas.d.ts`, at the project root, which stores the schema typings. The optional `--verbose` flag returns a detailed table of the generated schemas.
```

Reviewer: You've never heard of "displaing", @convly? It's a shortcut for displaying and explaining (or it's when you display something but with the accent from Marseille. 😂 (thanks for pointing it ❤️ )

> To use `content-types:generate-types`run the following code in a terminal at the project root:

Reviewer 2: `ts:generate-types`

<code-group>
<code-block title="NPM">

```sh

> npm run strapi content-types:generate-types --verbose #optional flag

Reviewer 2: ts:generate-types

```

</code-block>

<code-block title="YARN">

```sh

> yarn strapi content-types:generate-types --verbose #optional flag

Reviewer 2: `ts:generate-types`

- Run `yarn build` or `npm run build` in the plugin directory `src/admin/plugins/[my-plugin-name]` to build the admin panel including the plugin.

::: note
It is not necessary to repeat the `yarn` or `npm install` command after the initial installation. The `yarn build` or `npm run build` command is necessary to implement any plugin development that affects the admin panel.
:::

## Start Strapi programmatically

To start Strapi programmatically in a TypeScript project the Strapi instance requires the compiled code location. This section describes how to set and indicate the compiled code directory.

### Use the `strapi()` factory

Strapi can be run programmatically by using the `strapi()` factory. Since the code of TypeScript projects is compiled in a specific directory, the parameter `distDir` should be passed to the factory to indicate where the compiled code should be read:

```js
// path: ./src/plugins/<plugin-name>/server/index.js 

const strapi = require('@strapi/strapi');

const app = await strapi({ distDir: './dist' });
```

### Use the `strapi.compile()` function

> The `strapi.compile()` function should be mostly used for developing tools that need to start a Strapi instance and detect whether the project includes TypeScript code. `strapi.compile()` automatically detects the project language. If the project code contains any TypeScript code, `strapi.compile()` compiles the code and returns a context with specific values for the directories that Strapi requires:

Reviewer 2: Perfect!

## Start Strapi programmatically

> Starting Strapi programmatically in a TypeScript project requires passing the output directory where the code is compiled.

Reviewer 2: passing where?

```suggestion
Starting Strapi programmatically in a TypeScript project requires indicating to Strapi where the code is compiled.
```

> ### Use the `strapi()` function

Reviewer 2: ```suggestion
### Use the `strapi()` factory
```

> Strapi can be run programmatically by calling the `strapi()` function. Since the code of TypeScript projects is compiled in a specific directory, the parameter `distDir` should be passed to the `strapi()` function to indicate where the compiled code should be read:

Reviewer 2: ```suggestion
Strapi can be run programmatically by using the `strapi()` factory. Since the code of TypeScript projects is compiled in a specific directory, the parameter `distDir` should be passed to the factory to indicate where the compiled code should be read:
```

```js
// path: ./src/plugins/<plugin-name>/server/index.js 

const strapi = require('@strapi/strapi');

const app = await strapi({ distDir: './dist' });
```

### Use the `strapi.compile()` function

> The `strapi.compile()` method should be mostly used for developing tools that need to start a Strapi instance and detect whether the project includes TypeScript code. `strapi.compile()` automatically detects the project language. If the project code contains any TypeScript code, `strapi.compile()` compiles the code and returns a context with specific values for the directories that Strapi requires:

Reviewer 2: ```suggestion
The `strapi.compile()` function should be mostly used for developing tools that need to start a Strapi instance and detect whether the project includes TypeScript code. `strapi.compile()` automatically detects the project language. If the project code contains any TypeScript code, `strapi.compile()` compiles the code and returns a context with specific values for the directories that Strapi requires:
```
